### PR TITLE
refactor, build: Upstream fixes for the /crypto files.  Implement Keccak and SHA3

### PR DESCRIPTION
**Pull request by barton2526**
- Change CSipHasher's count variable to uint8_t
  Ref: https://github.com/bitcoin/bitcoin/pull/19931
> SipHash technically supports arbitrarily long inputs (at least, I couldn't find a limit in the paper), but only the low 8 bits of the length matter. Because of that we should use an unsigned type to track the length (as any signed type could overflow, which is UB). uint8_t is sufficient, however.
- Do not construct out-of-bound pointers in SHA2 code
  Ref: https://github.com/bitcoin/bitcoin/pull/15950
> This looks like an issue in the current SHA256/512 code, where a pointer outside of the area pointed to may be constructed (this is UB in theory, though in practice every supported platform treats pointers as integers).
- build: improve __builtin_clz* detection
  Ref: https://github.com/bitcoin/bitcoin/pull/19403
> The way we currently test for __builtin_clz* support with AC_CHECK_DECLS does not work with Clang
- Implement Keccak and SHA3_256
  Ref: https://github.com/bitcoin/bitcoin/pull/19841
> Add a simple (and initially unoptimized) Keccak/SHA3 implementation, as one will be needed for TORv3 support (the conversion from BIP155 encoding to .onion notation uses a SHA3-based checksum).

- Actually enable crypto libraries added in https://github.com/gridcoin-community/Gridcoin-Research/pull/1453

- compat: move cpuid 
- crypto: port some upstream changes

### CODE REVIEW

**Comment by barton2526** [comment]
Upcoming PR to add crypto_tests file depends on this PR due to https://github.com/gridcoin-community/Gridcoin-Research/pull/2365/commits/795e59c1034710c9cc24aeec6738b72726dbdce2

File: src/Makefile.am [hunk 1]
```diff
@@ -323,14 +323,22 @@ crypto_libgridcoin_crypto_base_a_SOURCES = \
...
     crypto/ripemd160.cpp \
     crypto/ripemd160.h \
     crypto/sha1.cpp \
     crypto/sha1.h \
     crypto/sha256.cpp \
     crypto/sha256.h \
+    crypto/sha256_avx2.cpp \
+    crypto/sha256_shani.cpp \
+    crypto/sha256_sse4.cpp \
+    crypto/sha256_sse41.cpp \
```
Comment (jamescowens) [inline src/Makefile.am]: I think these are covered below?

Comment (barton2526) [reply src/Makefile.am]: fixed

**Comment by barton2526** [comment]
Updated with 2 commits from #2368

**Review by jamescowens — COMMENTED** [review]
Are there tests for sha3?

**Review by jamescowens — APPROVED** [review]
utACK

**Comment by barton2526** [comment]
> 
> 
> Are there tests for sha3?

These are in the upcoming PR which will add the entire crypto_tests file

https://github.com/gridcoin-community/Gridcoin-Research/pull/2365#issuecomment-949027246

**Pull request merged**